### ISSUE TITLE
Converting Chart.js HTML to PDF crashes at step 1

### ISSUE THREAD

**Issue by ppluijten** [wkhtmltopdf/wkhtmltopdf#3551]
When converting a HTML page containing Chart.js to PDF, the tool crashes (without an exception) at Step 1. For input I used: [http://www.chartjs.org/samples/latest/charts/pie.html](http://www.chartjs.org/samples/latest/charts/pie.html)
Is there any way I can use this type of chart in the PDF conversion?

**Comment by daudmalik06** [comment]
@ppluijten  you can also try  this new library which uses chrome as browser so no more these kind of issues , very simple interface give it a try
https://github.com/daudmalik06/PhpChromeToPdf

**Comment by ppluijten** [comment]
Thank you for the tip, I will try it out for personal purposes.

Unfortunately for the current use case I am running it in an Azure WebJob, which has no user interface. The WkHtmlToPdf tool is perfect for this because it can be run as a console tool from the WebJob without requiring any front-end or interfacing tools or libraries.

A work-around I found is running a quick javascript-script before the PDF conversion which turns all 'canvas' objects into 'img' tags:

```
$("canvas").each(function () {
    // Create an Image object to capture the DataURI of the canvas
    var image = new Image();
    image.src = this.toDataURL("image/png");

    // Copy all styling attributes to take into account the current height
    $(image).attr("style", $(this).attr("style"));
    $(image).attr("width", $(this).attr("width"));
    $(image).attr("height", $(this).attr("height"));

    // Replace the canvas object with its corresponding img-tag
    $(this).replaceWith($(image));
});
```

**Comment by daudmalik06** [comment]
@ppluijten  ahh cool solution ,
 but the library i am mentioning is also command line based 
i didn't mean gui library :) 
have a look ,

**Comment by Avorathol** [comment]
Try the bleeding edge version (0.13.0-alpha), this one works with latest ChartJS.

**Comment by ppluijten** [comment]
Thank you for the tip, @Avorathol. Although using the 0.13.0-alpha version no longer results in an exceptionless crash, unfortunately it refuses to create the PDF:

![image](https://user-images.githubusercontent.com/975661/28461154-9af5d2f8-6e15-11e7-9ee8-70897d8020c2.png)

It ends with `QPainter::begin(): Returned false` and does not create a PDF file.

**Comment by ppluijten** [comment]
Oh, this error is not related to the input page, even when using https://www.google.com as input (which works with the stable version) results in the same behaviour.

**Comment by ppluijten** [comment]
Even with a `--javascript-delay` of 2000 (2 seconds) the page remains empty except for the buttons at the bottom.

**Comment by ppluijten** [comment]
@daudmalik06, the PhpChromeToPdf library assumes you have Chrome installed on your computer; because I am trying to run this on a Microsoft Azure WebJob, I do not have this available.

**Comment by Avorathol** [comment]
The `<div>` including the `<canvas>` needs a fix size to make it work, which isn't the case on ChartJS online exemples.
Get the html locally, change JS scripts url conveniently, and change the div by `<div id="canvas-holder" style="width: 510px; height: 510px;">`

**Comment by daudmalik06** [comment]
@ppluijten can't you download a binary ?

**Comment by ppluijten** [comment]
@daudmalik06, I have not tried that yet, that might be an option. Will try that soon.

**Comment by ppluijten** [comment]
@Avorathol, that makes sense; will try to update my HTML and see what happens.

**Comment by VitaliyMF** [comment]
It seems latest wkhtmltopdf release (0.12.4) has an regression that causes crush without errors when chartjs is used:

```
> wkhtmltopdf.exe --javascript-delay 1000 http://www.chartjs.org/samples/latest/charts/bar/vertical.html test.pdf 
```
The same command works fine (without errors, PDF is produced) for 0.12.3.2 (it is empty because of wrapping div is missed to workaround this known issue).

**Comment by charuhasrocks** [comment]
@Avorathol after including div it also results in empty pdf.currently i m using stable version 0.12.4 for wkhtlmltopdf but not working in my case.
<html>
    <head>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
    <script src = 'https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.7.1/Chart.bundle.min.js' ></script>
    </head>
    <body>
    <div style="width: 510px; height: 510px;">
    <canvas id="canvas" height="300" width="500"></canvas>
    <img id="imgWrap"></img>
    </div>
    </body>
    </html>
    <script>
    $(document).ready(function () {
            var ctx = document.getElementById('canvas').getContext("2d");
            //var d = new Chart(ctx, config);
            var myLineChart = new Chart(ctx, {
                type: 'line',
                data: {
                    labels: ["60", "40"],
                    datasets: [{               
                    data: ["67.8721351268983", "25.7067271673383"],
                }]
                },
                options : {
                animation: false
                }
            });
            $("canvas").each(function () {
                // Create an Image object to capture the DataURI of the canvas
                var image = new Image();
                image.src = this.toDataURL("image/png");
    
                // Copy all styling attributes to take into account the current height
                $(image).attr("style", $(this).attr("style"));
                $(image).attr("width", $(this).attr("width"));
                $(image).attr("height", $(this).attr("height"));
    
                // Replace the canvas object with its corresponding img-tag
                $(this).replaceWith($(image));
        });
    });
    </script>
